atc/db/worker_resource_cache.go

```diff
@@ -70,54 +81,61 @@ func (workerResourceCache WorkerResourceCache) FindOrCreate(tx Tx) (*UsedWorkerR
 	}, nil
 }
 
+// Find looks for a worker resource cache by resource cache id and worker name. It may find multiple
+// worker resource caches, just return the first valid one.
+// Note: Find doesn't consider workerResourceCache.SourceWorker as a search condition.
 func (workerResourceCache WorkerResourceCache) Find(runner sq.Runner) (*UsedWorkerResourceCache, bool, error) {
-	baseResourceType := workerResourceCache.ResourceCache.BaseResourceType()
-	usedWorkerBaseResourceType, found, err := WorkerBaseResourceType{
-		Name:       baseResourceType.Name,
-		WorkerName: workerResourceCache.WorkerName,
-	}.Find(runner)
-	if err != nil {
-		return nil, false, err
-	}
-
-	if !found {
-		return nil, false, nil
-	}
-
-	id, found, err := workerResourceCache.find(runner, usedWorkerBaseResourceType)
+	ids, workerBasedResourceTypeIds, found, err := workerResourceCache.find(runner, workerResourceCache.WorkerName)
 	if err != nil {
 		return nil, false, err
 	}
 
 	if found {
-		return &UsedWorkerResourceCache{
-			ID: id,
-		}, true, nil
+		// Verify worker base resource type
+		for i, workerBasedResourceTypeId := range workerBasedResourceTypeIds {
+			_, foundBrt, err := WorkerBaseResourceType{}.FindById(runner, workerBasedResourceTypeId)
+			if err != nil {
+				return nil, false, err
+			}
+			if foundBrt {
+				return &UsedWorkerResourceCache{ID: ids[i]}, true, nil
+			}
+		}
 	}
 
 	return nil, false, nil
 }
 
-func (workerResourceCache WorkerResourceCache) find(runner sq.Runner, usedWorkerBaseResourceType *UsedWorkerBaseResourceType) (int, bool, error) {
-	var id int
+func (workerResourceCache WorkerResourceCache) find(runner sq.Runner, workerName string) ([]int, []int, bool, error) {
+	var ids, workerBasedResourceTypeIds []int
 
-	err := psql.Select("id").
+	rows, err := psql.Select("id, worker_base_resource_type_id").
 		From("worker_resource_caches").
 		Where(sq.Eq{
-			"resource_cache_id":            workerResourceCache.ResourceCache.ID(),
-			"worker_base_resource_type_id": usedWorkerBaseResourceType.ID,
+			"resource_cache_id": workerResourceCache.ResourceCache.ID(),
+			"worker_name":       workerName,
 		}).
 		Suffix("FOR SHARE").
 		RunWith(runner).
-		QueryRow().
-		Scan(&id)
+		Query()
 	if err != nil {
-		if err == sql.ErrNoRows {
-			return 0, false, nil
+		return nil, nil, false, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var id, workerBasedResourceTypeId int
+		err := rows.Scan(&id, &workerBasedResourceTypeId)
+		if err != nil {
+			return nil, nil, false, err
 		}
+		ids = append(ids, id)
+		workerBasedResourceTypeIds = append(workerBasedResourceTypeIds, workerBasedResourceTypeId)
+	}
 
-		return 0, false, err
+	if len(workerBasedResourceTypeIds) == 0 {
```

Comment:
Is there a reason that we need this check? 🤔  If the length of them is 0, wouldn't that mean that the value of `ids` and `workerBasedResourceTypeIds` will be nil as well (because there would be no rows fetched from the query)?

Reply:
Yes, `ids` and `workerBasedResourceTypeIds` are nil, but I want the third returned value to be `false`.